app.js

```diff
@@ -148,6 +160,66 @@ app.get('/macros/:macro.json', function (req, res) {
   }
 });
 
+// List all devices in JSON format
+app.get('/devices.json', function(req, res) {
+    // TODO: Should return a nicely formatted response, not just the config object
+    res.json(config.devices);
+});
+
+// List all commands for :device in JSON format
+app.get('/devices/:device.json', function(req, res) {
+    if (config.devices && config.devices[req.params.device]) {
+        // TODO: Should return a nicely formatted response, not just the config object
+        res.json(config.devices[req.params.device]);
+    } else {
+        res.send(404);
+    }
+});
+
+// Get device state
+app.get('/devices/:device', function(req, res) {
+    if (config.devices &&
+        config.devices[req.params.device] &&
+        config.devices[req.params.device]['state']) {
+
+        var state = config.devices[req.params.device]['state'];
+
+        // Build request to make to get device state
+        var stateReq = {
+          method: state.method || "GET",
+          url: state.url,
+          form: state.body
+        };
+
+        // Make request for state, include body of response in response
+        request(stateReq, function (error, response, body) {
+          // TODO: How to parse the response and only return relevant portion?
+          var bodyJson = JSON.parse(body);
+          res.json(bodyJson);
+        });
+
+    } else {
+        // Device doesn't have a state, return 404
+        res.send(404);
+    }
+});
+
+// Execute device action
+app.post('/devices/:device/:command', function(req, res) {
```

Comment:
I need some clarity when talking about device - in the past hours I used
the Term "device" for 'lirc', 'gpio' or 'zigbee' endpoints from the macro
engine perspective.

If I got it right, you are using 'device' for a single device like the
receiver. In case of the Denon Receiver I own, I really want to use the IR
to change its settings, but i can fetch its state via web interface. The
interface is very slow when changing its state.

Bottomline: I think it is possible if not probable that we'll see IR
devices with physical state...

On Sun, Jan 17, 2016 at 11:20 PM, Alex Bain <notifications@github.com>
wrote:

> In app.js
> <https://github.com/alexbain/lirc_web/pull/36#discussion_r49953537>:
>
> > +
> > +        // Make request for state, include body of response in response
> > +        request(stateReq, function (error, response, body) {
> > +          // TODO: How to parse the response and only return relevant portion?
> > +          var bodyJson = JSON.parse(body);
> > +          res.json(bodyJson);
> > +        });
> > +
> > +    } else {
> > +        // Device doesn't have a state, return 404
> > +        res.send(404);
> > +    }
> > +});
> > +
> > +// Execute device action
> > +app.post('/devices/:device/:command', function(req, res) {
>
> I was trying to define what an abstract device would need and settled on
> two types of functionality:
>
>    - Getting state
>    - Triggering commands
>
> So you could "GET" the device to trigger requesting the current state, or
> you could POST to an endpoint to trigger a command.
>
> IR devices wouldn't have state, though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/alexbain/lirc_web/pull/36/files#r49953537>.
>